File: backend/pkg/database/repository/image.go
package repository

import (
	"context"
	"encoding/json"
	"time"

	"backend/pkg/core"
	"backend/pkg/database/models"
	"backend/pkg/types"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// ImageRepository defines the interface for image generation data access
type ImageRepository interface {
	FindByID(ctx context.Context, id uuid.UUID) (*types.ImageGeneration, error)
	FindByRequestID(ctx context.Context, requestID string) (*types.ImageGeneration, error)
	Save(ctx context.Context, img *types.ImageGeneration) error
	Update(ctx context.Context, img *types.ImageGeneration) error
}

// imageRepository provides data access for image generations
type imageRepository struct {
	db *gorm.DB
}

// NewImageRepository creates a new ImageRepository
func NewImageRepository(db *gorm.DB) ImageRepository {
	return &imageRepository{db: db}
}

// imageModelToType converts a database model to types
func imageModelToType(m *models.ImageGeneration) *types.ImageGeneration {
	var metaData map[string]interface{}
	if m.MetaData != nil {
		_ = json.Unmarshal(m.MetaData, &metaData)
	}

	var completedAt *time.Time
	if m.CompletedAt != nil {
		t, err := time.Parse(time.RFC3339, *m.CompletedAt)
		if err == nil {
			completedAt = &t
		}
	}

	return &types.ImageGeneration{
		ID:           m.ID,
		Prompt:       m.Prompt,
		Provider:     m.Provider,
		ModelName:    m.ModelName,
		RequestID:    m.RequestID,
		Status:       m.Status,
		OutputURL:    m.OutputURL,
		FileIndexID:  m.FileIndexID,
		ErrorMessage: m.ErrorMessage,
		MetaData:     metaData,
		CreatedAt:    m.CreatedAt,
		CompletedAt:  completedAt,
	}
}

// imageTypeToModel converts a types type to database model
func imageTypeToModel(i *types.ImageGeneration) *models.ImageGeneration {
	var metaData datatypes.JSON
	if i.MetaData != nil {
		data, _ := json.Marshal(i.MetaData)
		metaData = datatypes.JSON(data)
	}

	var completedAt *string
	if i.CompletedAt != nil {
		s := i.CompletedAt.Format(time.RFC3339)
		completedAt = &s
	}

	return &models.ImageGeneration{
		ID:           i.ID,
		Prompt:       i.Prompt,
		Provider:     i.Provider,
		ModelName:    i.ModelName,
		RequestID:    i.RequestID,
		Status:       i.Status,
		OutputURL:    i.OutputURL,
		FileIndexID:  i.FileIndexID,
		ErrorMessage: i.ErrorMessage,
		MetaData:     metaData,
		CreatedAt:    i.CreatedAt,
		CompletedAt:  completedAt,
	}
}

// FindByID retrieves an image generation by its ID
func (r *imageRepository) FindByID(ctx context.Context, id uuid.UUID) (*types.ImageGeneration, error) {
	var model models.ImageGeneration
	if err := r.db.WithContext(ctx).First(&model, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return imageModelToType(&model), nil
}

// FindByRequestID retrieves an image generation by its request ID
func (r *imageRepository) FindByRequestID(ctx context.Context, requestID string) (*types.ImageGeneration, error) {
	var model models.ImageGeneration
	if err := r.db.WithContext(ctx).Where("request_id = ?", requestID).First(&model).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return imageModelToType(&model), nil
}

// Save creates a new image generation
func (r *imageRepository) Save(ctx context.Context, img *types.ImageGeneration) error {
	model := imageTypeToModel(img)
	if model.ID == uuid.Nil {
		model.ID = uuid.New()
		img.ID = model.ID
	}
	return r.db.WithContext(ctx).Create(model).Error
}

// Update updates an existing image generation
func (r *imageRepository) Update(ctx context.Context, img *types.ImageGeneration) error {
	model := imageTypeToModel(img)
	return r.db.WithContext(ctx).Save(model).Error
}

File: backend/src/database/repository/image.rs
use async_trait::async_trait;
use diesel::{ExpressionMethods, OptionalExtension, QueryDsl, SelectableHelper};
use diesel_async::RunQueryDsl;
use serde_json::Value;
use uuid::Uuid;

use crate::{
    core::image::{IMAGE_STATUS_PENDING, ImageGeneration, ImageRepository},
    database::{
        models::image::{ImageChangeset, ImageRow, NewImageRow},
        pool::PgPool,
    },
    error::AppError,
    schema::imagen_request,
};

#[derive(Clone)]
pub struct DieselImageRepository {
    pool: PgPool,
}

impl DieselImageRepository {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    async fn connection(
        &self,
    ) -> Result<
        diesel_async::pooled_connection::deadpool::Object<diesel_async::AsyncPgConnection>,
        AppError,
    > {
        self.pool.get().await.map_err(|_| AppError::Database)
    }
}

#[async_trait]
impl ImageRepository for DieselImageRepository {
    async fn find_by_id(&self, id: Uuid) -> Result<ImageGeneration, AppError> {
        let mut connection = self.connection().await?;
        imagen_request::table
            .find(id)
            .select(ImageRow::as_select())
            .first(&mut connection)
            .await
            .optional()
            .map_err(|_| AppError::Database)?
            .ok_or(AppError::NotFound)
            .map(Into::into)
    }

    async fn find_by_request_id(&self, request_id: &str) -> Result<ImageGeneration, AppError> {
        let mut connection = self.connection().await?;
        imagen_request::table
            .filter(imagen_request::request_id.eq(request_id))
            .select(ImageRow::as_select())
            .first(&mut connection)
            .await
            .optional()
            .map_err(|_| AppError::Database)?
            .ok_or(AppError::NotFound)
            .map(Into::into)
    }

    async fn save(&self, image: &mut ImageGeneration) -> Result<(), AppError> {
        if image.id.is_nil() {
            image.id = Uuid::new_v4();
        }
        let mut connection = self.connection().await?;
        diesel::insert_into(imagen_request::table)
            .values(new_row(image))
            .execute(&mut connection)
            .await
            .map(|_| ())
            .map_err(|_| AppError::Database)
    }

    async fn update(&self, image: &ImageGeneration) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        let affected = diesel::update(imagen_request::table.find(image.id))
            .set(changeset(image))
            .execute(&mut connection)
            .await
            .map_err(|_| AppError::Database)?;
        if affected == 0 {
            diesel::insert_into(imagen_request::table)
                .values(new_row(image))
                .execute(&mut connection)
                .await
                .map_err(|_| AppError::Database)?;
        }
        Ok(())
    }
}

impl From<ImageRow> for ImageGeneration {
    fn from(row: ImageRow) -> Self {
        Self {
            id: row.id,
            prompt: row.prompt,
            provider: row.provider,
            model_name: row.model_name,
            request_id: row.request_id.unwrap_or_default(),
            status: row
                .status
                .unwrap_or_else(|| IMAGE_STATUS_PENDING.to_owned()),
            output_url: row.output_url.unwrap_or_default(),
            file_index_id: row.file_index_id,
            error_message: row.error_message.unwrap_or_default(),
            meta_data: row.meta_data.and_then(|value| match value {
                Value::Object(map) => Some(map.into_iter().collect()),
                _ => None,
            }),
            created_at: row.created_at,
            completed_at: row.completed_at,
        }
    }
}

fn new_row(image: &ImageGeneration) -> NewImageRow {
    NewImageRow {
        id: image.id,
        prompt: image.prompt.clone(),
        provider: image.provider.clone(),
        model_name: image.model_name.clone(),
        request_id: Some(image.request_id.clone()),
        status: image.status.clone(),
        output_url: Some(image.output_url.clone()),
        file_index_id: image.file_index_id,
        error_message: Some(image.error_message.clone()),
        meta_data: metadata_value(image),
        created_at: image.created_at.unwrap_or_else(chrono::Utc::now),
        completed_at: image.completed_at,
    }
}

fn changeset(image: &ImageGeneration) -> ImageChangeset {
    ImageChangeset {
        prompt: image.prompt.clone(),
        provider: image.provider.clone(),
        model_name: image.model_name.clone(),
        request_id: Some(image.request_id.clone()),
        status: Some(image.status.clone()),
        output_url: Some(image.output_url.clone()),
        file_index_id: image.file_index_id,
        error_message: Some(image.error_message.clone()),
        meta_data: image
            .meta_data
            .as_ref()
            .map(|map| Value::Object(map.clone().into_iter().collect())),
        completed_at: image.completed_at,
    }
}

fn metadata_value(image: &ImageGeneration) -> Value {
    Value::Object(
        image
            .meta_data
            .clone()
            .map(IntoIterator::into_iter)
            .map(Iterator::collect)
            .unwrap_or_default(),
    )
}
